jetbrains-core/src/software/aws/toolkits/jetbrains/services/caws/CawsHttpAuthProvider.kt

```diff
@@ -3,14 +3,44 @@
 
 package software.aws.toolkits.jetbrains.services.caws
 
+import com.intellij.openapi.application.ModalityState
 import com.intellij.openapi.project.Project
+import com.intellij.openapi.ui.MessageDialogBuilder
 import com.intellij.util.AuthData
 import git4idea.remote.GitHttpAuthDataProvider
+import software.aws.toolkits.jetbrains.core.awsClient
+import software.aws.toolkits.jetbrains.core.credentials.sono.SonoCredentialManager
+import software.aws.toolkits.jetbrains.services.caws.pat.generateAndStorePat
 import software.aws.toolkits.jetbrains.services.caws.pat.getPat
+import software.aws.toolkits.jetbrains.utils.computeOnEdt
+import software.aws.toolkits.resources.message
+import java.time.Instant
+import java.util.concurrent.atomic.AtomicLong
 
 class CawsHttpAuthProvider : GitHttpAuthDataProvider {
+    // globally only offer to make a new PAT at most once every 15s
+    private val lastRefreshPrompt = AtomicLong(0)
+
+    // framework will actually call this twice on first reject
     override fun forgetPassword(project: Project, url: String, authData: AuthData) {
-        super.forgetPassword(project, url, authData)
+        synchronized(this) {
+            val now = Instant.now().epochSecond
+            if (now - lastRefreshPrompt.getAndSet(now) < 15) {
```

Comment:
Should there be a user warning here? Should this be logged?

Reply:
it's a hack around framework behavior if nothing changes after this method call